main.py

```diff
@@ -567,12 +571,12 @@ def handle_interface_up(self, event):
 
     @listen_to('.*.switch.interface.created')
     def on_interface_created(self, event):
-        """Update the topology based on a Port Create event."""
-        self.handle_interface_created(event)
+        """Update the topology based on a Port Create event.
 
-    def handle_interface_created(self, event):
-        """Update the topology based on a Port Create event."""
-        self.handle_interface_up(event)
+        It's handled as a link_up in case a switch send a
+        created event again and it can be belong to a link.
+        """
+        self.handle_link_up(event.content['interface'])
```

Comment:
Vinicius, it looks like after applying this change (i.e., handling an interface up as link up) one of the end-to-end tests started failing:

```
_______ TestE2ETopology.test_090_removing_interfaces_metadata_persistent _______

self = <tests.test_e2e_05_topology.TestE2ETopology instance at 0x7fe5959a3908>

    def test_090_removing_interfaces_metadata_persistent(self):
        """
        Test /api/kytos/topology/v3/interfaces/{interface_id}/metadata/{key} on DELETE
        supported by:
            /api/kytos/topology/v3/interfaces/{interface_id}/metadata on POST
            and
            /api/kytos/topology/v3/interfaces/{interface_id}/metadata on GET
        """
        # It fails due to a bug, reported to Kytos team

        interface_id = "00:00:00:00:00:00:00:01:4"

        # Insert interface metadata
        payload = {"tmp_key": "tmp_value"}
        key = next(iter(payload))

        api_url = KYTOS_API + '/topology/v3/interfaces/%s/metadata' % interface_id
        response = requests.post(api_url, data=json.dumps(payload), headers={'Content-type': 'application/json'})
        assert response.status_code == 201

        self.restart()

        # Verify that the metadata is inserted
        api_url = KYTOS_API + '/topology/v3/interfaces/%s/metadata' % interface_id
        response = requests.get(api_url)
        data = response.json()
        keys = data['metadata'].keys()
>       assert key in keys
E       AssertionError: assert 'tmp_key' in []

tests/test_e2e_05_topology.py:361: AssertionError
```

In the function `handle_interface_up()` after activating the interface it always calls `update_instance_metadata()`. The function `handle_link_up()` on the other hand may terminate without doing the same step ([when the interface does not have a link](https://github.com/kytos-ng/topology/blob/fix/issue_34/main.py#L633-L634), for instance -- besides other scenarios).

I must admit I didn't get why handling the port_created as link_up instead of interface_up. Can you please check again this point?

Reply:
Just trying to contribute a bit to the above situation, if I revert the change on handling port created the end-to-end test stops failing and the behavior after reconnection remains the goal: interface became active again. I hope it helps.

```
diff --git a/main.py b/main.py
index 4325b9c..b132913 100644
--- a/main.py
+++ b/main.py
@@ -571,12 +571,12 @@ class Main(KytosNApp):  # pylint: disable=too-many-public-methods

     @listen_to('.*.switch.interface.created')
     def on_interface_created(self, event):
-        """Update the topology based on a Port Create event.
+        """Update the topology based on a Port Create event."""
+        self.handle_interface_created(event)

-        It's handled as a link_up in case a switch send a
-        created event again and it can be belong to a link.
-        """
-        self.handle_link_up(event.content['interface'])
+    def handle_interface_created(self, event):
+        """Update the topology based on a Port Create event."""
+        self.handle_interface_up(event)

     def handle_interface_down(self, event):
         """Update the topology based on a Port Modify event.
@@ -627,7 +627,6 @@ class Main(KytosNApp):  # pylint: disable=too-many-public-methods

     def handle_link_up(self, interface):
         """Notify a link is up."""
-        interface.activate()
         with self._links_lock:
             link = self._get_link_from_interface(interface)
         if not link:
@@ -636,6 +635,7 @@ class Main(KytosNApp):  # pylint: disable=too-many-public-methods
             other_interface = link.endpoint_b
         else:
             other_interface = link.endpoint_a
+        interface.activate()
         if other_interface.is_active() is False:
             return
         if link.is_active() is False:
```

Reply:
@italovalcy, indeed this end-to-end test has caught a bug (nice), Let's fix this. Let me answer your questions first:

> I must admit I didn't get why handling the port_created as link_up instead of interface_up. Can you please check again this point?

It's related to this point here https://github.com/kytos-ng/topology/pull/38#discussion_r774201621, it turns out that after a hankshake, currently, once `get_switch_or_create` is called on `of_core` when handling the hanskshake, we have this [`set_switch_options` method](https://github.com/kytos-ng/kytos/blob/master/kytos/core/controller.py#L668) that is creating new interfaces, so at the moment we can have an existing interface being created again (maybe this might be a point for improvement later on `kytos`?), that can lead to cases where an existing link will have new object references for these interfaces, so `created` events can bring a link up for instance, in a case where a switch reboots, I'm seeing his behavior on mininet. Since interface `created` can bring existing link ups, and other NApps like [`mef_eline` are subscribers of `"kytos/topology.link_(down|up)"`](https://github.com/kytos-ng/mef_eline/blob/master/main.py#L621) we need to make sure to also fire that event off, otherwise it could leave margin for subscribers to miss this event that can cause other issues. 

Here's a log simulating a reboot for instance (I added some temporary log.info entries `xyz handling interface.create` instead of debugging to minimize the number of messages to facilitate showing this here):

```
2022-01-12 13:58:17,987 - INFO [kytos.napps.kytos/of_core] (ThreadPoolExecutor-0_4) Connection ('127.0.0.1', 34398), Switch 00:00:00:00:00:00:00:01: OPENFLOW HANDSHAKE COMPLETE
2022-01-12 13:58:17,999 - INFO [kytos.napps.kytos/of_core] (ThreadPoolExecutor-0_1) Connection ('127.0.0.1', 34402), Switch 00:00:00:00:00:00:00:02: OPENFLOW HANDSHAKE COMPLETE
2022-01-12 13:58:18,000 - INFO [kytos.napps.kytos/flow_manager] (Thread-39) Send FlowMod from request dpid: 00:00:00:00:00:00:00:01, command: add, force: False, flows_dict: {'flows': [
{'priority': 1000, 'table_id': 0, 'match': {'dl_type': 35020, 'dl_vlan': 3799}, 'actions': [{'action_type': 'output', 'port': 4294967293}]}]}
2022-01-12 13:58:18,003 - INFO [kytos.napps.kytos/topology] (ThreadPoolExecutor-0_9) xyz handling interface.create {'interface': Interface('s1', 4294967294, Switch('00:00:00:00:00:00:0
0:01'))}
2022-01-12 13:58:18,012 - INFO [kytos.napps.kytos/topology] (ThreadPoolExecutor-0_10) xyz handling interface.create {'interface': Interface('s1-eth2', 2, Switch('00:00:00:00:00:00:00:0
1'))}
2022-01-12 13:58:18,015 - INFO [kytos.napps.kytos/topology] (ThreadPoolExecutor-0_5) xyz handling interface.create {'interface': Interface('s1-eth1', 1, Switch('00:00:00:00:00:00:00:01
'))}
2022-01-12 13:58:18,020 - INFO [kytos.napps.kytos/flow_manager] (ThreadPoolExecutor-0_6) Flow saved in kytos.flow.persistence.365f042b363b4b4fad68c6876641c184
2022-01-12 13:58:18,026 - INFO [kytos.napps.kytos/flow_manager] (Thread-41) Send FlowMod from request dpid: 00:00:00:00:00:00:00:02, command: add, force: False, flows_dict: {'flows': [
{'priority': 1000, 'table_id': 0, 'match': {'dl_type': 35020, 'dl_vlan': 3799}, 'actions': [{'action_type': 'output', 'port': 4294967293}]}]}
2022-01-12 13:58:18,039 - INFO [kytos.napps.kytos/topology] (ThreadPoolExecutor-0_6) xyz handling interface.create {'interface': Interface('s2-eth1', 1, Switch('00:00:00:00:00:00:00:02
'))}
2022-01-12 13:58:18,039 - INFO [kytos.napps.kytos/of_core] (ThreadPoolExecutor-0_10) Connection ('127.0.0.1', 34400), Switch 00:00:00:00:00:00:00:03: OPENFLOW HANDSHAKE COMPLETE
2022-01-12 13:58:18,040 - INFO [kytos.napps.kytos/topology] (ThreadPoolExecutor-0_11) xyz handling interface.create {'interface': Interface('s2', 4294967294, Switch('00:00:00:00:00:00:
00:02'))}
2022-01-12 13:58:18,041 - INFO [kytos.napps.kytos/topology] (ThreadPoolExecutor-0_0) xyz handling interface.create {'interface': Interface('s2-eth2', 2, Switch('00:00:00:00:00:00:00:02
'))}
2022-01-12 13:58:18,041 - INFO [kytos.napps.kytos/topology] (ThreadPoolExecutor-0_8) xyz handling interface.create {'interface': Interface('s2-eth3', 3, Switch('00:00:00:00:00:00:00:02
'))}
```

Reply:
@italovalcy and on this commit here https://github.com/kytos-ng/topology/pull/38/commits/25e272d7e82f9cd238a21a4f1f3da724e00fe7ce, the `handle_interface_created` is also calling the `update_instance_metadata` so that should fix the issue you've caught. Let me know if you need more info or if it's still confusing. I really appreciated your help and inputs to fix this.

Reply:
@viniarck thank you for the detailed explanation! It makes sense to me. I'm gonna run a new subset of tests to validate the last commit you made